### PR TITLE
Surface API alternative PR

### PR DESCRIPTION
**Pull request by tomvanmele**
This PR is a refactored version of #1350 ...

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [ ] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [ ] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [ ] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [ ] I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by tomvanmele** [comment]
Some explanation...

* `compas.geometry.surfaces.surface.Surface` is like an ABC for parametric surfaces.
* `compas.geometry.surfaces.surface.Surface.__new__` has been replaced such that only subclasses can be instantiated.
* `__new__` re-overwrites of subclasses are therefore no longer necessary.
* `Surface` has many implementations (Conical, Cylindrical, Planar, Spherical, Toroidal, Nurbs). only `NurbsSurface` requires a plugin implementation.
* `from_native` has been added to `Surface` since CAD engines can provide a "native" surface from which all parametric functionality can be derived, even without knowing what kind of surface it is.
* all specific factory methods have been moved to `NurbsSurface` since they indeed produce a Nurbs surface.
* elementary surfaces can be instantiated directly from the specific pure Python classes (`ConicalSurface`, `CylindricalSurface`, ...)
* a few small bugs have been fixed along the way.
* similar changes should be made to the Curve API...

all of this has been tested with OCC as plugin and using the viewer for visualisation. Rhino should still be verified.

File: src/compas/geometry/surfaces/nurbs.py [hunk 1]
```diff
@@ -109,25 +144,46 @@ def __data__(self):
...
+            knots_v,
+            mults_u,
+            mults_v,
+            degree_u,
+            degree_v,
+            is_periodic_u,
+            is_periodic_v,
         )
 
+    # this may not really be necessary
```
Comment (chenkasirer) [inline src/compas/geometry/surfaces/nurbs.py]: shouldn't `NurbsSurface` be made ABC as well?

File: src/compas/geometry/surfaces/nurbs.py [hunk 2]
```diff
@@ -109,25 +144,46 @@ def __data__(self):
 
     @classmethod
     def __from_data__(cls, data):
+        """Construct a Nurbs surface from its data representation.
```
Comment (chenkasirer) [inline src/compas/geometry/surfaces/nurbs.py:147]: is the domain not part of the representation? can it somehow be inferred from another piece of data?

File: src/compas/geometry/surfaces/nurbs.py [hunk 3]
```diff
@@ -187,26 +244,155 @@ def degree_u(self):
...
+    def domain_v(self):
+        raise NotImplementedError
+
     # ==============================================================================
     # Constructors
     # ==============================================================================
 
+    @classmethod
+    def from_cylinder(cls, cylinder, *args, **kwargs):
+        """Construct a surface from a sphere.
```
Comment (chenkasirer) [inline src/compas/geometry/surfaces/nurbs.py]:
```suggestion
        """Construct a surface from a cylinder.
```

File: src/compas/geometry/surfaces/nurbs.py [hunk 4]
```diff
@@ -187,26 +244,155 @@ def degree_u(self):
...
+        Returns
+        -------
+        :class:`compas.geometry.NurbsSurface`
+
+        """
+        return nurbssurface_from_frame(cls, frame, *args, **kwargs)
+
+    @classmethod
+    def from_interpolation(cls, points, *args, **kwargs):
+        """Construct a surface from a frame.
```
Comment (chenkasirer) [inline src/compas/geometry/surfaces/nurbs.py:342]:
```suggestion
        """Construct a surface from an interpolation.
```

File: src/compas/geometry/surfaces/nurbs.py [hunk 5]
```diff
@@ -187,26 +244,155 @@ def degree_u(self):
...
             A CAD native surface object.
 
         Returns
         -------
         :class:`compas.geometry.NurbsSurface`
-            A COMPAS NURBS surface.
 
         """
-        return new_nurbssurface_from_native(cls, surface)
+        raise NotImplementedError
```
Comment (chenkasirer) [inline src/compas/geometry/surfaces/nurbs.py:395]: call plugin here?

File: src/compas/geometry/surfaces/surface.py [hunk 6]
```diff
@@ -51,7 +45,9 @@ class Surface(Geometry):
     """
 
     def __new__(cls, *args, **kwargs):
-        return new_surface(cls, *args, **kwargs)
+        if cls is Surface:
+            raise TypeError("Instantiating the base Surface class directly is not allowed.")
```
Comment (chenkasirer) [inline src/compas/geometry/surfaces/surface.py]: I would maybe change the wording a bit. the caller doesn't necessarily know an entirely different class is gonna be created, that's abstracted by the plugin system. if I saw something like this I'd go looking for the right child class to instantiate myself. Instead, they should know the default constructor is not allowed but the other ones yes.

**Review by chenkasirer — APPROVED** [review]
LGTM!

**Pull request merged**